diff --git a/SConstruct b/SConstruct
index 2281b8a77f..c0e8cc3fa6 100644
--- a/SConstruct
+++ b/SConstruct
@@ -543,6 +543,9 @@ if selected_platform in platform_list:
                 env.module_icons_paths.append(path + "/" + "icons")
             modules_enabled[name] = path
 
+            if getattr(config, "has_custom_iterator", False) and config.has_custom_iterator():
+                env.AppendUnique(CPPDEFINES=["CUSTOM_ITERATOR"])
+
         sys.path.remove(path)
         sys.modules.pop("config")
 
diff --git a/main/main.cpp b/main/main.cpp
index c492cfaad7..f029693444 100644
--- a/main/main.cpp
+++ b/main/main.cpp
@@ -1284,8 +1284,8 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 	}
 
 	/* todo restore
-    OS::get_singleton()->_allow_layered = GLOBAL_DEF("display/window/per_pixel_transparency/allowed", false);
-    video_mode.layered = GLOBAL_DEF("display/window/per_pixel_transparency/enabled", false);
+	OS::get_singleton()->_allow_layered = GLOBAL_DEF("display/window/per_pixel_transparency/allowed", false);
+	video_mode.layered = GLOBAL_DEF("display/window/per_pixel_transparency/enabled", false);
 */
 	GLOBAL_DEF("rendering/quality/intended_usage/framebuffer_allocation", 2);
 	GLOBAL_DEF("rendering/quality/intended_usage/framebuffer_allocation.mobile", 3);
@@ -2420,6 +2420,7 @@ bool Main::is_iterating() {
 	return iterating > 0;
 }
 
+#ifndef CUSTOM_ITERATOR
 // For performance metrics.
 static uint64_t physics_process_max = 0;
 static uint64_t idle_process_max = 0;
@@ -2582,6 +2583,7 @@ bool Main::iteration() {
 
 	return exit || auto_quit;
 }
+#endif // ~CUSTOM_ITERATOR
 
 void Main::force_redraw() {
 	force_redraw_requested = true;
